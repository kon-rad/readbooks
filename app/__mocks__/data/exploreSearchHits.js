
export const hits = [
  {
    id: "_uTRAwAAQBAJ",
    volumeInfo: {
      title: "Practical Modern JavaScript",
      imageLinks: {
        smallThumbnail: "http://books.google.com/books/content?id=9gUJAQAAMAAJ&printsec=frontcover&img=1&zoom=5&source=gbs_api"
      },
      publishedDate: "1995-12-01",
      authors: ['author first', 'author second']
    },
    searchInfo: {
      textSnippet: "Many webmasters don't have time to learn a high level programming language, "
    }
  },
  {
    id: "vmE4nBg4mpc",
    volumeInfo: {
      title: "Functional JavaScript",
      imageLinks: {
        smallThumbnail: "http://image.com"
      },
      publishedDate: "2009-12-01",
      authors: ['author first', 'author second']
    },
    searchInfo: {
      textSnippet: "Lorem ipsum"
    }
  }
];