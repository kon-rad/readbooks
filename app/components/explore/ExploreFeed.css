.feed {
    margin-top: 60px;
}

.feed__item {
    overflow: auto;
    min-width: 80px;
    cursor: pointer;
}

.feed__item:hover {
    cursor: pointer;
}

.feed__item_title {
    margin-top: 0;
    margin-bottom: 4px;
}
