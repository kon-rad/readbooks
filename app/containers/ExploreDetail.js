// @flow

import * as React from 'react';
import { Link } from 'react-router-dom';
import type { TitleResults } from '../types/BooksAPI';
import styles from './ExploreDetail.css';

type Props = {
  location: {
    state: TitleResults
  }
};

export default class ExploreDetail extends React.Component<Props> {

  addBook() {

  }

  render() {
    const imageUrl = this.props.location.state.volumeInfo.imageLinks
      ? this.props.location.state.volumeInfo.imageLinks.smallThumbnail
      : '/images/no_results.svg';
    const title = this.props.location.state
      ? this.props.location.state.volumeInfo.title
      : '';
    const publishedDate = this.props.location.state
      ? this.props.location.state.volumeInfo.publishedDate
      : '';

    let authors = '';
    if (this.props.location.state.volumeInfo.authors) {
      authors = this.props.location.state.volumeInfo.authors.join(', ');
    }

    return (
      <div className="segment">
        <div className="ui items">
          <div className="ui item">
            <div className="ui tiny-image">
              <img src={imageUrl} className={styles.detailImage} alt="" />
            </div>
            <div className={styles.detailContent}>
              <div className="content">
                <h1 className="ui header">{title}</h1>
                <Link to="/book/create" className="ui mini positive basic button">
                  Add Book
                </Link>
                <div className="meta">
                  <span className={styles.detailMeta}>Authors: {authors}</span>
                  <span className={styles.detailMeta}>
                    Published: {publishedDate}
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <h4 className="ui header">Description</h4>
        <p>{this.props.location.state.volumeInfo.description}</p>
      </div>
    );
  }
}
