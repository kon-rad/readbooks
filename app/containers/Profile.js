import * as React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import {
  NotificationContainer,
  NotificationManager
} from 'react-notifications';
import axios from 'axios';
import _ from 'lodash';
import Posts from '../components/profile/Posts';
import FollowingList from '../components/profile/FollowingList';
import FollowersList from '../components/profile/FollowersList';
import UserInfo from '../components/profile/UserInfo';
import * as actions from '../actions';

// Todo: Add maps, to allow debugging on compiled, browser code, (webpack)
const POSTS = 'posts';
const FOLLOWING = 'following';
const FOLLOWERS = 'followers';

export const TABS = {
  'posts': {
    id: 'posts',
    label: 'Posts'
  },
  'following': {
    id: 'following',
    label: 'Following'
  },
  'followers': {
    id: 'followers',
    label: 'Followers'
  }
};

class Profile extends React.Component {
  constructor(props: Props) {
    super(props);
    const { auth, match: { params: { id } } } = this.props;
    this.state = {
      activeTab: 'posts',
      id,
      following: auth ? auth.following : [],
      refreshFollowersList: false
    };
  }

  componentDidMount() {
    const { fetchProfile, match: { params: { id }}} = this.props;
    fetchProfile(id);
  }

  componentDidUpdate(prevProps) {
    const { auth } = this.props;
    if (auth !== prevProps.auth) {
      this.setState({ following: auth.following });
    }
  }

  followUser = async () => {
    const res = await axios.post(
      `/profile/${this.props.profile.user._id}/follow`
    );
    if (!res.data.error) {
      NotificationManager.success(
        `Success! Followed ${this.props.profile.user.username}`
      );
      this.setState({
        following: res.data.following,
        refreshFollowersList: !this.state.refreshFollowersList
      });
    }
  };

  unfollowUser = async () => {
    const res = await axios.post(
      `/profile/${this.props.profile.user._id}/unfollow`
    );
    if (!res.data.error) {
      NotificationManager.success(
        `Success! Unfollowed ${this.props.profile.user.username}`
      );
      this.setState({
        following: res.data.following,
        refreshFollowersList: !this.state.refreshFollowersList
      });
      this.forceUpdate();
    }
  };

  changeTab = tabId => {
    this.setState({
      activeTab: tabId
    });
  };

  renderButtons() {
    const { auth, profile } = this.props;
    const { following } = this.state;
    if (!auth) {
      return
    }
    if (auth._id === profile.user._id) {
      return (
        <div className="ui content">
          <a
            href={`/profile/${auth._id}/edit`}
            className="ui right floated mini primary button basic"
          >
            Edit
          </a>
          <a href="/logout" className="ui right floated mini button basic">
            Logout
          </a>
        </div>
      );
    }

    if (_.includes(following, profile.user._id)) {
      return (
        <div className="ui content">
          <button
            onClick={this.unfollowUser}
            className="ui right floated mini blue button basic"
          >
            Unfollow
          </button>
        </div>
      );
    }

    return (
      <div className="ui content">
        <button
          onClick={this.followUser}
          className="ui right floated mini primary button basic"
        >
          Follow
        </button>
      </div>
    );
  }

  renderTabs = () => {
    const { activeTab } = this.state;

    return Object.keys(TABS).map(tabId => {
      const tab = TABS[tabId];
      return(
        <a
          key={tabId}
          className={`item ${activeTab === tabId ? 'active' : ''}`}
          data-tab={tabId}
          onClick={() => this.changeTab(tabId)}
        >
          {tab.label}
        </a>
      );
    });
  };

  renderTabContent = () => {
    const { user: { _id } } = this.props.profile;
    const { activeTab, refreshFollowersList } = this.state;
    if (activeTab === POSTS) {
      return <Posts id={_id} />;
    }
    if (activeTab === FOLLOWING) {
      return <FollowingList id={_id} />
    }
    if (activeTab === FOLLOWERS) {
      return <FollowersList refresh={refreshFollowersList} id={_id} />
    }
  };

  render() {
    const { user } = this.props.profile;
    if (!user) {
      return <h3>Loading...</h3>;
    }

    const { activeTab } = this.state;
    const avatar = user.avatar
      ? '/' + user.avatar
      : '/images/avatar-placeholder.jpg';

    return (
      <div className="segment">
        {this.renderButtons()}
        <UserInfo user={user} avatar={avatar}/>
        <div className="ui top attached tabular menu">
          {this.renderTabs()}
        </div>
        <div
          key={activeTab}
          className={`ui bottom attached tab segment active`}
        >
          {this.renderTabContent()}
        </div>
        <NotificationContainer />
      </div>
    );
  }
}

function mapStateToProps(state) {
  return { auth: state.auth, profile: state.profile };
}

// using withRouter allows component to refresh when route param changes
export default withRouter(
  connect(
    mapStateToProps,
    actions
  )(Profile)
);
