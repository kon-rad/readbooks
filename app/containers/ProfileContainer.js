import * as React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import {
  NotificationContainer,
  NotificationManager
} from 'react-notifications';
import axios from 'axios';
import ProfileTabs from 'components/profile/ProfileTabs';
import ProfileTabContent from 'components/profile/ProfileTabContent';
import UserInfo from 'components/profile/UserInfo';
import * as actions from 'actions';

// todo: create tests for all code
// todo: add js linting

export class ProfileContainer extends React.Component {
  constructor(props: Props) {
    super(props);
    const { auth, match: { params: { id } } } = this.props;
    this.state = {
      activeTab: 'posts',
      id,
      following: auth ? auth.following : [],
      refreshFollowersList: false
    };
  }

  componentDidMount() {
    const { fetchProfile, match: { params: { id }}} = this.props;
    fetchProfile(id);
  }

  componentDidUpdate(prevProps) {
    const { auth } = this.props;
    if (auth !== prevProps.auth) {
      this.setState({ following: auth.following });
    }
  }

  followUser = async () => {
    const res = await axios.post(
      `/profile/${this.props.profile.user._id}/follow`
    );
    if (!res.data.error) {
      NotificationManager.success(
        `Success! Followed ${this.props.profile.user.username}`
      );
      this.setState({
        following: res.data.following,
        refreshFollowersList: !this.state.refreshFollowersList
      });
    }
  };

  unfollowUser = async () => {
    const res = await axios.post(
      `/profile/${this.props.profile.user._id}/unfollow`
    );
    if (!res.data.error) {
      NotificationManager.success(
        `Success! Unfollowed ${this.props.profile.user.username}`
      );
      this.setState({
        following: res.data.following,
        refreshFollowersList: !this.state.refreshFollowersList
      });
      this.forceUpdate();
    }
  };

  changeTab = tabId => {
    this.setState({
      activeTab: tabId
    });
  };

  renderButtons() {
    const { auth, profile: { user: { _id } } } = this.props;
    const { following } = this.state;
    if (!auth) {
      return;
    }

    if (auth._id === _id) {
      return (
        <div className="ui content">
          <a
            href={`/profile/${auth._id}/edit`}
            className="ui right floated mini primary button basic"
          >
            Edit
          </a>
          <a href="/logout" className="ui right floated mini button basic">
            Logout
          </a>
        </div>
      );
    }

    if (following.indexOf(_id) !== -1) {
      return (
        <div className="ui content">
          <button
            onClick={this.unfollowUser}
            className="ui right floated mini blue button basic"
          >
            Unfollow
          </button>
        </div>
      );
    }

    return (
      <div className="ui content">
        <button
          onClick={this.followUser}
          className="ui right floated mini primary button basic"
        >
          Follow
        </button>
      </div>
    );
  }

  render() {
    const { user } = this.props.profile;
    if (!user) {
      return <h3>Loading...</h3>;
    }

    const { activeTab, refreshFollowersList } = this.state;
    const avatar = user.avatar
      ? `/${user.avatar}`
      : '/images/avatar-placeholder.jpg';

    return (
      <div className="segment">
        {this.renderButtons()}
        <UserInfo user={user} avatar={avatar}/>
        <div className="ui top attached tabular menu">
          <ProfileTabs activeTab={activeTab} changeTab={this.changeTab}/>
        </div>
        <div
          key={activeTab}
          className="ui bottom attached tab segment active"
        >
          <ProfileTabContent
            activeTab={activeTab}
            refreshFollowersList={refreshFollowersList}
            userId={user._id}
          />
        </div>
        <NotificationContainer />
      </div>
    );
  }
}

function mapStateToProps(state) {
  return { auth: state.auth, profile: state.profile };
}

// using withRouter allows component to refresh when route param changes
export default withRouter(
  connect(
    mapStateToProps,
    actions
  )(ProfileContainer)
);
