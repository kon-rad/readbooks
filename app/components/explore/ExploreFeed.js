// @flow

import * as React from 'react';
import type { TitleResults } from 'app/types/BooksAPI';
import './ExploreFeed.css';
import { Link } from 'react-router-dom';

type Props = {
  searchHits: Array<TitleResults>
};

class ExploreFeed extends React.Component<Props> {

  constructor(props: Props) {
    super(props);
  }

  processPostFields = (hit) => {
    let thumbnail = hit.volumeInfo.imageLinks
      ? hit.volumeInfo.imageLinks.smallThumbnail
      : '/images/no_results.svg';
    let title = hit.volumeInfo.title;
    let publishedDate = hit.volumeInfo.publishedDate;
    let authors = hit.volumeInfo.authors ? hit.volumeInfo.authors.join(', ') : '';
    let description = hit.searchInfo ? hit.searchInfo.textSnippet : '';

    return { title, publishedDate, thumbnail, authors, description };
  };

  renderSearchHit = (hit: TitleResults): React.Element<'div'> => {
    let book = this.processPostFields(hit);
    const detail = {
      pathname: '/explore/' + hit.id,
      state: book
    };

    return (
      <div key={hit.id} className='feedItem ui segment'>
        <Link to={detail}>
          <div className="content">
            <div className="ui tiny left floated image">
              <img
                src={book.thumbnail}
                alt=""
              />
            </div>
            <div className="header">{book.title}</div>
            <div className="meta">{book.publishedDate}</div>
            <div
              className="description"
              dangerouslySetInnerHTML={{
                __html: book.description
              }}
            />
          </div>
        </Link>
      </div>
    );
  };

  render() {
    return (
      <div className='feedWrapper'>
        {this.props.searchHits.map(this.renderSearchHit)}
      </div>
    );
  }
}

export default ExploreFeed;
