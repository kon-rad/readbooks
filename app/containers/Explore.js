// @flow

import React from 'react';
import ExploreSearch from '../components/ExploreSearch';
import ExploreFeed from '../components/ExploreFeed';
import apiCaller from '../util/apiCaller';
import type { TitleResults } from '../types/BooksAPI';

type Props = {};

type State = {
  searchHits: Array<TitleResults>,
  query: string

};

class ExploreContainer extends React.Component<Props, State> {

  handleSearch = (query: string): void => {

    apiCaller(query)
      .then((res) => {
        this.setState({ searchHits: res.items || [] });
      })
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      searchHits: [],
      query: ''
    }
  }

  render() {
    return (
      <div>
        <ExploreSearch handleSearch={this.handleSearch} />
        <ExploreFeed searchHits={this.state.searchHits} />
      </div>
    )
  }
}

export default ExploreContainer;