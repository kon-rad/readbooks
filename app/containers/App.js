// @flow

import React from 'react';
import { Router, Route } from 'react-router-dom';
import ExploreContainer from './Explore';
import PostShow from '../components/post/PostShow';
import PostCreate from '../components/post/PostCreate';
import history from '../util/history';
import Profile from './Profile';
import { connect } from 'react-redux';
import * as actions from '../actions';
import 'react-notifications/lib/notifications.css';

type Props = {
  fetchUser: Function
};

class App extends React.Component<Props> {
  componentDidMount() {
    this.props.fetchUser();
  }

  render() {
    return (
      <Router history={history}>
        <div>
          <Route exact path="/explore/:id" component={PostShow} />
          <Route exact path="/explore" component={ExploreContainer} />
          <Route exact path="/post/create" component={PostCreate} />
          <Route
            exact
            path="/profile/:id"
            // this allows the component to refresh when the route param changes
            render={props => <Profile key={props.location.key} />}
          />
        </div>
      </Router>
    );
  }
}

export default connect(
  null,
  actions
)(App);
