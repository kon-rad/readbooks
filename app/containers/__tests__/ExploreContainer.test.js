import React from 'react';
import { shallow } from 'enzyme';
import { ExploreContainer } from 'containers/ExploreContainer';
import ExploreSearch from 'components/explore/ExploreSearch';
import ExploreFeed from 'components/explore/ExploreFeed';
import apiCaller from 'util/apiCaller';
import { hits } from 'components/__mocks__/data/exploreSearchHits';

jest.mock('util/apiCaller');

let wrapper;

describe('explore container', () => {

  beforeEach(() => {
    wrapper = shallow(
      <ExploreContainer />
    );
  });

  it('renders matching snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('shows a search component', () => {
    expect(wrapper.find(ExploreSearch).length).toEqual(1);
  });

  it('shows a search results feed component', () => {
    expect(wrapper.find(ExploreFeed).length).toEqual(1);
  });

  describe('the search functionality', () => {
    it('calls the bookSearch function when handleSearch is called', async () => {
      const data = { items: hits };
      const mockAPI = apiCaller.mockImplementation(() =>
        Promise.resolve(data)
      );

      wrapper.find(ExploreSearch).prop('handleSearch')('Searching for a book');
      await mockAPI();
      expect(mockAPI).toHaveBeenCalled();
    });

    it('initiates a book search when handleSearch is called', async () => {
      const data = { items: hits };
      const mockAPI = apiCaller.mockImplementation(() =>
        Promise.resolve(data)
      );

      wrapper.find(ExploreSearch).prop('handleSearch')('Searching for a book');
      await mockAPI();
      expect(wrapper.instance().state.searchHits.length).toEqual(2);
    });
  });
});
