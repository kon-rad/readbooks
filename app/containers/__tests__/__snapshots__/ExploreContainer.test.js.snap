// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`explore container renders matching snapshot 1`] = `
<div>
  <ExploreSearch
    handleSearch={[Function]}
  />
  <ExploreFeed
    searchHits={Array []}
  />
</div>
`;
